// twing/dist/cjs/lib/error/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createBaseError = void 0;
const createBaseError = (name, message, location, source, previous) => {
    const baseError = Error(message);
    baseError.name = name;
    const error = Object.create(baseError, {
        location: {
            get: () => location,
            set: (value) => {
                location = value;
                updateRepresentation();
            }
        },
        source: {
            get: () => source,
            set: (value) => {
                source = value;
                updateRepresentation();
            }
        },
        previous: {
            value: previous
        },
        rootMessage: {
            value: message
        },
        appendMessage: {
            value: (value) => {
                message += value;
                updateRepresentation();
            }
        }
    });
    const updateRepresentation = () => {
        let representation = message;
        let dot = false;
        if (representation.slice(-1) === '.') {
            representation = representation.slice(0, -1);
            dot = true;
        }
        let questionMark = false;
        if (representation.slice(-1) === '?') {
            representation = representation.slice(0, -1);
            questionMark = true;
        }
        if (source) {
            representation += ` in "${source}"`;
        }
        if (location !== undefined) {
            const { line, column } = location;
            representation += ` at line ${line}, column ${column}`;
        }
        if (dot) {
            representation += '.';
        }
        if (questionMark) {
            representation += '?';
        }
        baseError.message = representation;
    };
    updateRepresentation();
    Error.captureStackTrace(error, exports.createBaseError);
    return error;
};
exports.createBaseError = createBaseError;
